contracts/pools/aeth/StableSwapAETH.vy

```diff
@@ -0,0 +1,857 @@
+    raise
+
+
+@view
+@internal
+def get_D_mem(rates: uint256[N_COINS], _balances: uint256[N_COINS], amp: uint256) -> uint256:
+    result: uint256[N_COINS] = rates
+    for i in range(N_COINS):
+        result[i] = result[i] * _balances[i] / PRECISION
+    return self.get_D(result, amp)
+
+
+@view
+@external
+def get_virtual_price() -> uint256:
+    """
+    @notice The current virtual price of the pool LP token
+    @dev Useful for calculating profits
+    @return LP token virtual price normalized to 1e18
+    """
+    D: uint256 = self.get_D(self._xp(self._stored_rates()), self._A())
+    # D is in the units similar to DAI (e.g. converted to precision 1e18)
+    # When balanced, D = n * x_u - total virtual value of the portfolio
+    token_supply: uint256 = ERC20(self.lp_token).totalSupply()
+    return D * PRECISION / token_supply
+
+
+@view
+@external
+def calc_token_amount(amounts: uint256[N_COINS], is_deposit: bool) -> uint256:
+    """
+    @notice Calculate addition or reduction in token supply from a deposit or withdrawal
+    @dev This calculation accounts for slippage, but not fees.
+         Needed to prevent front-running, not for precise calculations!
+    @param amounts Amount of each coin being deposited
+    @param is_deposit set True for deposits, False for withdrawals
+    @return Expected amount of LP tokens received
+    """
+    amp: uint256 = self._A()
+    rates: uint256[N_COINS] = self._stored_rates()
+    _balances: uint256[N_COINS] = self.balances
+    D0: uint256 = self.get_D_mem(rates, _balances, amp)
+    for i in range(N_COINS):
+        _amount: uint256 = amounts[i]
+        if is_deposit:
+            _balances[i] += _amount
+        else:
+            _balances[i] -= _amount
+    D1: uint256 = self.get_D_mem(rates, _balances, amp)
+    token_amount: uint256 = ERC20(self.lp_token).totalSupply()
+    diff: uint256 = 0
+    if is_deposit:
+        diff = D1 - D0
+    else:
+        diff = D0 - D1
+    return diff * token_amount / D0
+
+@payable
+@external
+@nonreentrant('lock')
+def add_liquidity(amounts: uint256[N_COINS], min_mint_amount: uint256) -> uint256:
+    """
+    @notice Deposit coins into the pool
+    @param amounts List of amounts of coins to deposit
+    @param min_mint_amount Minimum amount of LP tokens to mint from the deposit
+    @return Amount of LP tokens received by depositing
+    """
+    assert not self.is_killed
+
+    rates: uint256[N_COINS] = self._stored_rates()
+    _lp_token: address = self.lp_token
+    token_supply: uint256 = ERC20(_lp_token).totalSupply()
+
+    # Initial invariant
+    amp: uint256 = self._A()
+    D0: uint256 = 0
+    old_balances: uint256[N_COINS] = self.balances
+    if token_supply != 0:
+        D0 = self.get_D_mem(rates, old_balances, amp)
+
+    new_balances: uint256[N_COINS] = old_balances
+    for i in range(N_COINS):
+        if token_supply == 0:
+            assert amounts[i] > 0
+        new_balances[i] = old_balances[i] + amounts[i]
```

Comment:
```suggestion
        new_balances[i] += amounts[i]
```

Use of `+=` in this case is more gas efficient because it results in slightly less use of memory.